client/X11/xf_rail.c

```diff
@@ -531,22 +548,337 @@ static BOOL xf_rail_window_delete(rdpContext* context,
+	if (!argbPixels)
+		goto error;
+
+	if (!convert_icon_color_to_argb(iconInfo, argbPixels))
+		goto error;
+
+	apply_icon_alpha_mask(iconInfo, argbPixels);
+
+	nelements = 2 + iconInfo->width * iconInfo->height;
+	pixels = realloc(railIcon->data, nelements * sizeof(*pixels));
+	if (!pixels)
+		goto error;
+
+	railIcon->data = pixels;
+	railIcon->length = nelements;
+
+	pixels[0] = iconInfo->width;
+	pixels[1] = iconInfo->height;
+
+	nextPixel = argbPixels;
+	for (i = 2; i < nelements; i++)
+	{
+		pixels[i] = read_color_quad(nextPixel);
+		nextPixel += 4;
+	}
+
+	free(argbPixels);
+
+	return TRUE;
+
+error:
+	free(argbPixels);
+	return FALSE;
+}
+
+static void xf_rail_set_window_icon(xfContext* xfc,
+                                    xfAppWindow* railWindow, xfRailIcon *icon,
+                                    BOOL replace)
+{
+	XChangeProperty(xfc->display, railWindow->handle, xfc->_NET_WM_ICON,
+		XA_CARDINAL, 32, replace ? PropModeReplace : PropModeAppend,
+		(unsigned char*) icon->data, icon->length);
+	XFlush(xfc->display);
+}
+
+static xfAppWindow* xf_rail_get_window_by_id(xfContext* xfc, UINT32 windowId)
+{
+	return (xfAppWindow*) HashTable_GetItemValue(xfc->railWindows,
+		(void*)(UINT_PTR) windowId);
+}
+
 static BOOL xf_rail_window_icon(rdpContext* context,
                                 WINDOW_ORDER_INFO* orderInfo, WINDOW_ICON_ORDER* windowIcon)
 {
 	xfContext* xfc = (xfContext*) context;
-	xfAppWindow* railWindow = (xfAppWindow*) HashTable_GetItemValue(xfc->railWindows,
-	                          (void*)(UINT_PTR) orderInfo->windowId);
+	xfAppWindow* railWindow;
+	xfRailIcon *icon;
+	BOOL replaceIcon;
 
+	railWindow = xf_rail_get_window_by_id(xfc, orderInfo->windowId);
 	if (!railWindow)
+		return TRUE;
+
+	icon = RailIconCache_Lookup(xfc->railIconCache,
+		windowIcon->iconInfo->cacheId,
+		windowIcon->iconInfo->cacheEntry);
+	if (!icon)
+	{
+		WLog_DBG(TAG, "failed to get icon from cache %02X:%04X",
+			windowIcon->iconInfo->cacheId,
+			windowIcon->iconInfo->cacheEntry);
 		return FALSE;
+	}
+
+	if (!convert_rail_icon(windowIcon->iconInfo, icon))
+	{
+		WLog_DBG(TAG, "failed to convert icon for window %08X", orderInfo->windowId);
+	}
+
+	replaceIcon = !!(orderInfo->fieldFlags & WINDOW_ORDER_STATE_NEW);
+	xf_rail_set_window_icon(xfc, railWindow, icon, replaceIcon);
 
 	return TRUE;
 }
 
 static BOOL xf_rail_window_cached_icon(rdpContext* context,
                                        WINDOW_ORDER_INFO* orderInfo, WINDOW_CACHED_ICON_ORDER* windowCachedIcon)
 {
+	xfContext* xfc = (xfContext*) context;
+	xfAppWindow* railWindow;
+	xfRailIcon *icon;
+	BOOL replaceIcon;
+
+	railWindow = xf_rail_get_window_by_id(xfc, orderInfo->windowId);
+	if (!railWindow)
+		return TRUE;
+
+	icon = RailIconCache_Lookup(xfc->railIconCache,
+		windowCachedIcon->cachedIcon.cacheId,
+		windowCachedIcon->cachedIcon.cacheEntry);
+	if (!icon)
+	{
+		WLog_DBG(TAG, "failed to get icon from cache %02X:%04X",
```

Comment:
It should not happen at all because the spec says that the server MUST send us correct IDs. So yeah, it makes sense to print a warning here. (I don't think error is appropriate as the remote session is not going to terminate soon because of this.)